### PR TITLE
Migrate to post 0.1.80 API, new Async model

### PR DESCRIPTION
**Pull request by FlorianBorn71**

### CODE REVIEW

File: ArrtModel/Model/ArrFrontend.cpp [hunk 1]
```diff
@@ -42,60 +42,52 @@ void ArrFrontend::reconnectAccount()
...
-            m_sessionPropertiesAsync->Completed([thisPtr, frontend](const RR::ApiHandle<RR::SessionPropertiesArrayAsync>& /*async*/) {
-                {
-                    std::unique_lock<std::mutex> lk(thisPtr->m_mutex);
-                    thisPtr->m_condVar.notify_one();
-                }
-                QMetaObject::invokeMethod(QApplication::instance(),
-                                          [thisPtr, frontend]() {
-                                              if (thisPtr != nullptr)
+        m_sessionPropertiesAsync = RR::Status::InProgress;
+        client->GetCurrentRenderingSessionsAsync([thisPtr, client](RR::Status status, RR::ApiHandle<RR::RenderingSessionPropertiesArrayResult> result) {
```
Comment (mafranc) [inline ArrtModel/Model/ArrFrontend.cpp:68]: Out of curiosity, what happens if there was a previous async call that is not completed yet? And is there a difference from the previous behaviour (was it cancelled if the async object was deleted?)

File: ArrtModel/Model/ArrFrontend.h [hunk 2]
```diff
@@ -36,9 +36,9 @@ class ArrFrontend : public QObject
     std::string m_accountKey;
     std::string m_accountDomain;
 
-    RR::ApiHandle<RR::AzureFrontend> m_rrFrontend;
-    RR::ApiHandle<RR::SessionPropertiesArrayAsync> m_sessionPropertiesAsync;
-
+    RR::ApiHandle<RR::RemoteRenderingClient> m_rrClient;
+    RR::Status m_sessionPropertiesAsync = RR::Status::OK;
```
Comment (mafranc) [inline ArrtModel/Model/ArrFrontend.h:40]: maybe m_sessionPropertiesAsyncStatus

File: ArrtModel/Model/ArrServiceStats.cpp [hunk 3]
```diff
@@ -27,43 +27,37 @@ void ArrServiceStats::updateStats(RR::ApiHandle<RR::AzureSession> session)
...
-            m_currentStats.m_memoryCPU.addValue(m_lastPerformanceAssessment.memoryCPU.aggregate, m_tick);
-            m_currentStats.m_memoryGPU.addValue(m_lastPerformanceAssessment.memoryGPU.aggregate, m_tick);
-            m_currentStats.m_networkLatency.addValue(m_lastPerformanceAssessment.networkLatency.aggregate, m_tick);
-            m_currentStats.m_polygonsRendered.addValue(m_lastPerformanceAssessment.polygonsRendered.aggregate, m_tick);
-            m_runningPerformanceAssesment = {};
+            qWarning(LoggingCategory::renderingSession) << tr("Error retrieving the performance assessment. Failure reason: ") << m_runningPerformanceAssessment.getStatus();
         }
         else
         {
-            qWarning(LoggingCategory::renderingSession) << tr("what?");
```
Comment (mafranc) [inline ArrtModel/Model/ArrServiceStats.cpp:66]: Ahah what was this?

File: ArrtModel/ViewModel/ApplicationModel.cpp [hunk 4]
```diff
@@ -26,7 +26,7 @@
 #pragma clang diagnostic ignored "-Wunused-parameter"
 #pragma clang diagnostic ignored "-Wsign-compare"
 #pragma warning(push)
-#pragma warning(disable : 4100)
+#pragma warning(disable : 4100 4189)
```
Comment (mafranc) [inline ArrtModel/ViewModel/ApplicationModel.cpp:29]: which one is 4189?

File: ArrtModel/Model/ConversionManager.h [hunk 5]
```diff
@@ -120,9 +120,9 @@ struct Conversion
             return QCoreApplication::tr("Conversion %1").arg(id);
         }
     }
-
-    RR::ApiHandle<RR::ConversionStatusAsync> m_statusAsync = nullptr;
-    RR::ApiHandle<RR::StartAssetConversionAsync> m_conversionCall = nullptr;
+    std::atomic_bool m_statusQueryInProgress = false;
+    //RR::ApiHandle<RR::ConversionStatusAsync> m_statusAsync = nullptr;
```
Comment (mafranc) [inline ArrtModel/Model/ConversionManager.h:124]: remove if not needed

File: ArrtModel/Model/ArrAsync.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+// Specialization of ArrAsyncStatus that wraps around session->Connection()->QueryServerPerformanceAssessmentAsync
+class ArrPerformanceAssessmentAsync : public ArrAsyncStatus<RR::PerformanceAssessment>
+{
+public:
+    // Performs the query asynchronously
+    void queryServerPerformanceAssessmentAsync(RR::ApiHandle<RR::RenderingSession> session)
+    {
+        m_isCompleted = false;
+        m_isRunning = true;
+        m_status = RR::Status::InProgress;
```
[the remark this answers is not in the document]
Comment (mafranc) [reply ArrtModel/Model/ArrAsync.h:66]: Actually is this the only class using this? I thought you wanted to apply it to the most of the async operations.

**Review by mafranc — CHANGES_REQUESTED** [review]

File: ArrtModel/Model/ArrAsync.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (mafranc) [reply ArrtModel/Model/ArrAsync.h:66]: Consider maybe getting rid of it them

File: ArrtModel/Model/ArrFrontend.h [hunk 7]
```diff
@@ -1,13 +1,13 @@
 #pragma once
 #include <Model/IncludesAzureRemoteRendering.h>
+#include <Model/ArrAsync.h>
 #include <Model/Settings/AccountConnectionStatus.h>
```
[the remark this answers is not in the document]
Comment (FlorianBorn71) [reply ArrtModel/Model/ArrFrontend.h:4]: did it now (and will do it for final commit)

File: ArrtModel/Model/ArrSessionManager.cpp [hunk 8]
```diff
@@ -83,18 +78,18 @@ namespace
...
         sessionInfo[QLatin1String("vm_size")] = toString(info.Size);
         return logger << QCoreApplication::tr("Create rendering session info:") << PrettyJson(sessionInfo);
     }
 
-    inline QDebug& operator<<(QDebug& logger, const RR::RenderingSessionUpdateParams& info)
+    inline QDebug& operator<<(QDebug& logger, const RR::RenderingSessionUpdateOptions& info)
     {
         QJsonObject sessionInfo;
-        sessionInfo[QLatin1String("max_lease")] = QString("%1:%2:%3").arg(info.MaxLease.hour).arg(info.MaxLease.minute).arg(info.MaxLease.second);
+        sessionInfo[QLatin1String("max_lease")] = QString("%1:%2").arg(info.MaxLeaseInMinutes / 60).arg(info.MaxLeaseInMinutes);
```
[the remark this answers is not in the document]
Comment (FlorianBorn71) [reply ArrtModel/Model/ArrSessionManager.cpp:92]: oups ;-)

File: ArrtModel/Model/ArrSessionManager.cpp [hunk 9]
```diff
@@ -187,15 +182,36 @@ ArrSessionManager::ArrSessionManager(ArrFrontend* frontEnd, Configuration* confi
...
+                                RR::ApiHandle<RR::RenderingSession> session;
+                                std::string errorStr;
+                                if (status == RR::Status::OK)
+                                {
+                                    session = result->GetSession();
+                                    errorStr = result->GetContext().ErrorMessage;
+                                }
+                                else
+                                {
+                                    errorStr = "General failure";
```
[the remark this answers is not in the document]
Comment (FlorianBorn71) [reply ArrtModel/Model/ArrSessionManager.cpp:197]: done

File: ArrtModel/Model/ArrServiceStats.cpp (hunk 3 above)
Comment (FlorianBorn71) [reply ArrtModel/Model/ArrServiceStats.cpp:66]: an impossible fallback ;-)

File: ArrtModel/ViewModel/ApplicationModel.cpp (hunk 4 above)
Comment (FlorianBorn71) [reply ArrtModel/ViewModel/ApplicationModel.cpp:29]: unreferenced local variable, unfortunately the RemoteRendering.inl is full of that. It's auto-generated code, so it cannot easily be removed, but there is already an ApiGen ticket for that

File: ArrtModel/Model/ArrFrontend.cpp (hunk 1 above)
Comment (FlorianBorn71) [reply ArrtModel/Model/ArrFrontend.cpp:68]: The condition has not changed at all compared to previous version

File: ArrtModel/Model/ArrAsync.h [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    // Indicates whether operation has been completed with an error.
+    inline bool isFaulted() const
+    {
+        return m_status != RR::Status::OK;
+    }
+
+protected:
+    void setResult(RR::Status status, const T& res)
+    {
+        m_status = status;
```
[the remark this answers is not in the document]
Comment (FlorianBorn71) [reply ArrtModel/Model/ArrAsync.h]: removed the whole wrapper

File: ArrtModel/Model/ArrAsync.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (FlorianBorn71) [reply ArrtModel/Model/ArrAsync.h]: removed wrapper

**Review by mafranc — APPROVED** [review]

**Pull request merged**